src/output/plugins/visualization/SoundInfoCache.hxx

```diff
@@ -0,0 +1,115 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef SOUND_INFO_CACHE_HXX_INCLUDED
+#define SOUND_INFO_CACHE_HXX_INCLUDED
+
+#include "output/Timer.hxx"
+#include "pcm/AudioFormat.hxx"
+#include "thread/Mutex.hxx"
+
+#include <chrono>
+#include <memory>
+
+/**
+ * \brief Thread-safe cache for recent PCM data
+ *
+ *
+ * Class SoundInfoCache maintains a ring buffer (AKA circular buffer) for PCM
+ * data to cap the amount of of memory used. It keeps two pointers into that
+ * buffer: the beginning and the end of valid data, along with the timestamps
+ * corresponding to each.
+ *
+ * The general contract is that once the ctor returns, the caller has an
+ * instance with an empty ring buffer & that is ready to accept data. Time starts
+ * from the first invocation of add(). Successive invocations of add() are
+ * assumed to represent contiguous ranges of sound data (i.e. there is no way
+ * to represent gaps).
+ *
+ * Instances may have their methods invoked by multiple threads, so any method
+ * invocation will block on acquiring a Mutex. I had initially considered a
+ * single-writer, multi-reader lock in the interests of allowing many
+ * simultaneous reads, but in practice it would not be an improvement, since
+ * there is only one reader & one writer, and the writer, empirically, is the
+ * more frequent caller.
+ *
+ * A circular buffer is surprisingly difficult to write. I considered
+ * abstracting this implementation into a general purpose library class, but
+ * there are a number of implementation-specific choices arguing against that:
+ * 
+ * - using a flag versus wasting a slot to distinguish full from empty
+ * - overwrite versus drop when new data won't fit
+ * - copy in bulk (via `mempcy()`) versus copying slot-by-slot
+ * 
+ * In the end I decided to just write an application-specific implementation.
+ *
+ *
+ */
+
+class SoundInfoCache {
+public:
+	typedef std::chrono::steady_clock::duration Duration;
+	typedef std::chrono::microseconds Time;
+
+private:
+	AudioFormat fmt;
+	/// Time per frame, in seconds
+	double secs_per_frame;
+	/// Sample size, in bytes
+	unsigned frame_size;
+	// Mutex guarding the ring buffer since instances will be accessed from
+	// multiple threads
+	mutable Mutex mutex;
+	/// ring buffer size/capacity, in bytes
+	std::size_t cb_ring;
+	/// this is the ring buffer
+	std::unique_ptr<uint8_t[]> ring;
+	/// # of bytes currently in the ring buffer (as distinct from capacity)
+	std::size_t cb;
+	/// Valid PCM data exists in buf[p0, p1)
+	ptrdiff_t p0, p1;
```

Comment:
Why is this a signed integer?

Reply:
No good reason; must have been an artifact of a first implementation. Changed, in any event.